File: javascript/easy/plusOne.js
/**
 * @param {number[]} digits
 * @return {number[]}
 */
var plusOne = function(digits) {
    let carry = plusOneRecursive(0, digits), i = 0;
    if(carry > 0) {
        digits.unshift(carry);
    }
    while(i < digits.length && digits[i] == 0) {
        i++;
    }
    return digits.slice(i, digits.length);
};

var plusOneRecursive = function(i, digits) {
    if(i == digits.length) return 1;
    let sum = digits[i] + plusOneRecursive(i+1, digits);
    digits[i] = (sum%10);
    return parseInt(sum/10);
};

File: ruby/easy/plus_one.rb
# @param {Integer[]} digits
# @return {Integer[]}
def plus_one(digits)
    i, j, carry = digits.length-1, 0, 1
    while i >= 0
        sum = digits[i] + carry
        digits[i] = (sum % 10)
        carry = (sum/10).to_i
        i -= 1
    end
    digits.unshift(carry) if carry > 0
    while j < digits.length && digits[j] == 0
        j += 1
    end
    return digits.drop(j)
end
